src/database.ts

```diff
@@ -0,0 +1,221 @@
+/**
+ * Promise database implementation, with stricter typing.
+ * By Mia
+ * @author mia-pi-git
+ */
+import * as mysql from 'mysql2';
+import SQL, {SQLStatement} from 'sql-template-strings';
+import {Config} from './config-loader';
+
+export type SQLInput = string | number | null;
+export interface ResultRow {[k: string]: SQLInput}
+
+export const databases: PSDatabase[] = [];
+
+export class PSDatabase {
+	pool: mysql.Pool;
+	prefix: string;
+	constructor(config: {[k: string]: any}) {
+		config = {...config};
+		this.prefix = config.prefix || "";
+		if (config.prefix) {
+			delete config.prefix;
+		}
+		this.pool = mysql.createPool(config);
+		if (!databases.includes(this)) databases.push(this);
+	}
+	query<T = ResultRow>(query: SQLStatement) {
+		const err = new Error();
+		return new Promise<T[]>((resolve, reject) => {
+			// this cast is safe since it's only an array of
+			// arrays if we specify it in the config.
+			// we do not do that and it is not really useful for any of our cases.
+			this.pool.query(query.sql, query.values, (e, results: mysql.RowDataPacket[]) => {
```

Comment:
I see no reason for this weird error constuction or the callback. I use mysql on the smogon monorepo and the driver supports promises natively

Reply:
We had an issue with `mysql` - I think it didn't support UNIX sockets? Whatever it was, it made mysql2 necessary.